patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java

```diff
@@ -0,0 +1,192 @@
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.gson.Gson;
+import org.objectweb.asm.Type;
+
+import net.fabricmc.loader.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
+
+public class ModFileScanData {
+	private Runnable initFunc;
+	private AnnotationStorage annotationStorage;
+	private Set<AnnotationData> annotationData;
+
+	public ModFileScanData(String modid) {
+		initFunc = ()
+				-> {
+			ModContainer modContainer = FabricLoader.INSTANCE.getModContainer(modid)
+					.orElseThrow(
+							() -> new RuntimeException("Cannot get mod container for " + modid)
+					);
+			Path annotationJsonPath = modContainer.getPath("annotations.json");
+
+			try {
+				InputStream outputStream = Files.newInputStream(annotationJsonPath);
+				Gson gson = new Gson();
+				annotationStorage = gson.fromJson(
+						new InputStreamReader(outputStream), AnnotationStorage.class
+				);
+				annotationData = annotationStorage.entries.stream()
+						.map(entry -> getAnnotationData(entry))
+						.collect(Collectors.toSet());
+			} catch (FileNotFoundException e) {
+				e.printStackTrace();
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		};
+	}
+
+	private void initIfNeeded() {
+		if (initFunc != null) {
+			initFunc.run();
+			initFunc = null;
+		}
+	}
+
+	public Set<AnnotationData> getAnnotations() {
+		initIfNeeded();
+		return annotationData;
+	}
+
+	private static AnnotationData getAnnotationData(AnnotationStorage.Entry entry) {
+		Type annotationType = Type.getType(entry.annotationType);
+		Type targetInType = Type.getType("L" + entry.targetInClass + ";");
+		return new AnnotationData(
+				annotationType,
+				entry.targetType,
+				targetInType,
+				entry.target
+		);
+	}
+
+	public static class AnnotationData {
+		private final Type annotationType;
+		private final ElementType targetType;
+		private final Type clazz;
+		private final String memberName;
+
+		//lazy evaluated
+		private Map<String, Object> annotationData;
+
+		public AnnotationData(
+				final Type annotationType,
+				final ElementType targetType,
+				final Type clazz,
+				final String memberName
+		) {
+			this.annotationType = annotationType;
+			this.targetType = targetType;
+			this.clazz = clazz;
+			this.memberName = memberName;
+		}
+
+		public Type getAnnotationType() {
+			return annotationType;
+		}
+
+		public ElementType getTargetType() {
+			return targetType;
+		}
+
+		public Type getClassType() {
+			return clazz;
+		}
+
+		public String getMemberName() {
+			return memberName;
+		}
+
+		public Map<String, Object> getAnnotationData() {
+			if (annotationData == null) {
+				initAnnotationData();
+			}
+
+			return annotationData;
+		}
+
+		private void initAnnotationData() {
+			try {
+				Class<?> clazzObj = Class.forName(clazz.getClassName());
```

Comment:
```suggestion
				// TODO: This *may* load classes in the wrong order, but it shouldn't be an issue
				Class<?> clazzObj = Class.forName(clazz.getClassName());
```

Reply:
This shouldn't load classes in the first place

Reply:
Not doing so would require serializing the annotation properties into the JSON which could be complex. In fact, I suggested just reading it using reflection instead, which does have this issue. Though, as you've pointed out, that approach could be problematic.